clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -257,8 +257,13 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, C
                 }
             }
         }
-        if (!expiredBatches.isEmpty())
+        if (!expiredBatches.isEmpty()) {
             log.trace("Expired {} batches in accumulator", count);
+            for (RecordBatch batch : expiredBatches) {
+                batch.expire();
+                deallocate(batch);
```

Comment:
@rajinisivaram, it's safe to `deallocate` without a lock. My concern was the `expire` method. In the `abortBatches` case, we `close` the batch with the lock held and invoke `done` outside the lock. Maybe we should do the same here to make it easier to reason about?